expose a hook to skip tables during iteration

benesch opened the pull request:
As discussed on the mailing list (["Skipping entire SSTs while iterating"](https://groups.google.com/forum/#!topic/rocksdb/ujHCJVLrHlU)), this patch adds a `table_filter` to `ReadOptions` that allows specifying a callback to be executed during iteration before each table in the database is scanned. The callback is passed the table's properties; the table is scanned iff the callback returns true.

This can be used in conjunction with a `TablePropertiesCollector` to dramatically speed up scans by skipping tables that are known to contain irrelevant data for the scan at hand.

We're using this [downstream in CockroachDB](https://github.com/cockroachdb/cockroach/blob/master/pkg/storage/engine/db.cc#L2009-L2022) already. With this feature, under ideal conditions, we can reduce the time of an incremental backup in  from hours to seconds.

FYI, the first commit in this PR fixes a segfault that I unfortunately have not figured out how to reproduce outside of CockroachDB. I'm hoping you accept it on the grounds that it is not correct to return 8-byte aligned memory from a call to `malloc` on some 64-bit platforms; one correct approach is to infer the necessary alignment from `std::max_align_t`, as done here. As noted in the first commit message, the bug is tickled by having a`std::function` in `struct ReadOptions`. That is, the following patch alone is enough to cause RocksDB to segfault when run from CockroachDB on Darwin.

```diff
--- a/include/rocksdb/options.h
+++ b/include/rocksdb/options.h
@@ -1546,6 +1546,13 @@ struct ReadOptions {
   // Default: false
   bool ignore_range_deletions;
 
+  // A callback to determine whether relevant keys for this scan exist in a
+  // given table based on the table's properties. The callback is passed the
+  // properties of each table during iteration. If the callback returns false,
+  // the table will not be scanned.
+  // Default: empty (every table will be scanned)
+  std::function<bool(const TableProperties&)> table_filter;
+
   ReadOptions();
   ReadOptions(bool cksum, bool cache);
 };
```

/cc @danhhz

yiwu-arbug commented:
@benesch Thank you for your contribution! Can you send the first as a separate PR? I think it should be a separate fix. Also can you give more explanation of the segfault (stacktrace maybe)? Thanks!

@siying @IslamAbdelRahman Any comment on the rest of the PR? Thanks.

IslamAbdelRahman commented:
This is a really good idea
PR LGTM, but maybe we should make it more explicit in the comments that this option is only effective for Iterators since ReadOptions is also used with point lookups

benesch commented:
Thanks for the review! I've opened #2347 to track the misalignment in the arena allocator, and updated the comment on the `table_filter` option to explicitly call out it has no impact on point lookups.

benesch commented:
Heh, it appears the Travis test failures here are a reproduction of the very issue I claimed I couldn't repro in #2347. I'll rebase this as soon as #2347 lands.

siying commented:
Can you rebase?

ajkr commented:
cc @mikhail-antonov - this may be useful for your incremental replication.

sagar0 commented:
@benesch Could you please rebase?
We'd like to accept this PR, but is #2347 really a pre-requisite for this one? There has been a considerable interest around this PR as it could be useful in various scenarios.

benesch commented:
Ok, rebased! Would also love to get this merged. Afraid that #2347 is absolutely necessary, though; otherwise, the `std::function` I added in `options.h` causes a segfault because the existing arena only returns 8-byte aligned memory, not 16-byte aligned memory. You can see the particular test that fails on Travis.

benesch commented:
The long-standing test failure in https://github.com/facebook/rocksdb/pull/2347 is finally resolved, so I don't think there any remaining blockers!

sagar0 commented:
Thanks for merging this @yiwu-arbug . 
@mikhail-antonov -- you can use this in #2999 now :)

mikhail-antonov commented:
Thanks guys! Glad to see it landed.

I definitely can make use of it.
